### PR TITLE
Revert "Revert "Add date opened to every update block""

### PR DESCRIPTION
**Pull request by czue**
Reverts dimagi/commcare-hq#12538

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/case/mock.py [hunk 1]
```diff
@@ -137,6 +137,7 @@ def __init__(self,
             })
         self['update'].update({
             'external_id':              external_id,
+            'date_opened':              date_opened,
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/mock.py]: E241 multiple spaces after ':'

File: corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py [hunk 2]
```diff
@@ -106,11 +107,13 @@ def test_couch_soft_rebuild(self):
         self.assertEqual(case.p1, 'p1-1') # original
         self.assertEqual(case.p2, 'p2-2') # updated
         self.assertEqual(case.p3, 'p3-2') # new
-        self.assertEqual(3, len(case.actions)) # create + 2 updates
-        a1 = case.actions[1]
+        self.assertEqual(4, len(case.actions)) # create + update (2 actions) + 2 updates
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py]: E261 at least two spaces before inline comment

**Comment by benrudolph** [comment]
he's going for it!

**Comment by mkangia** [comment]
@proteusvacuum : Added case's date_opened for phone backup but one addition was at a different location than you did. Plus like suggested used date instead of date time.
Pushing the changes to check the build on travis. Would go ahead and then fix them.

Post changes in phone restore response have :
1)  for version=2.0 => `<case xmlns="http://commcarehq.org/case/transaction/v2" case_id="awesomeness" date_modified="2016-08-02T13:27:30.768000Z" date_opened="2016-08-02" user_id="maybe_another_awesomeness">`

2) for version 1.0 => 
`<case>
    <case_id>f8e10160-132f-45b6-ac14-f6d2cd9aa35e</case_id>
    <date_modified>2016-08-02T15:52:34.057000Z</date_modified>
    <date_opened>2016-08-02</date_opened>
    <create>
      <case_type_id>case</case_type_id>
      <user_id>79dff4c745d271e58e8c8d646c1ac107</user_id>
      <case_name>yaha</case_name>
      <external_id/>
    </create>
    <update>
      <owner_id>79dff4c745d271e58e8c8d646c1ac107</owner_id>
      <sample_choice_question>choice3</sample_choice_question>
      <sample_number_question>3</sample_number_question>
    </update>
  </case>`

**Comment by proteusvacuum** [comment]
Hey @mkangia ,
I think we want to leave the `date_opened` item out of the case attributes for now since the [spec](https://github.com/dimagi/commcare-core/wiki/casexml20#case-action-elements) calls for it as a node in the update block. i.e. we want something that looks like:
```xml
<case>
    <update>
        <date_opened>2016-08-01</date_opened>
    </update>
</case>
```
I think the change you need to make is to change [this line](https://github.com/dimagi/commcare-hq/pull/12690/files#diff-d9b6db2d90177dd837136a58fca27673R161) to only add the `date_opened` as a date without a time:
```py
if self.case.opened_on:
    element.append(safe_element("date_opened", json_format_date(self.case.opened_on)))
```
You might also need to update references to `CaseBlock` so that the `date_modified` parameter only takes a date, e.g. [this line](https://github.com/dimagi/commcare-hq/pull/12690/files#diff-50e3d7daf1818bae5de7a28f6e71d0c3R104) should be something like
```py
CaseBlock(
    ...
    date_modified=now,
    date_opened=now.date(),
)
```
This will probably make all the tests that I added in my original PR complain, so you'll also have to fix those.
Hope that helps as a starting point!

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 3]
```diff
@@ -29,7 +29,16 @@ def __init__(self, xml_element, version=V2):
         self.version = version
 
     def to_string(self):
-        return ElementTree.tostring(self.xml_element)
+        restore_element_copy = ElementTree.fromstring(ElementTree.tostring(self.xml_element))
+        # hack need to explicitly convert date_opened from datetimes to dates in restores
+        # because the mobile can't handle datetimes.
+        # see: http://manage.dimagi.com/default.asp?230043 for more info
+        date_opened_element = restore_element_copy.find(
+            '{http://commcarehq.org/case/transaction/v2}update/{http://commcarehq.org/case/transaction/v2}date_opened'
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: E501 line too long (118 > 115 characters)

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 4]
```diff
@@ -29,7 +29,16 @@ def __init__(self, xml_element, version=V2):
...
+        )
+        if date_opened_element is not None and date_opened_element.text:
+            date_opened_element.text = date_opened_element.text[:10]
```
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: Is this because all the tests initialize the date_opened as a date time? I think you should just be able to change this whenever a `CaseBlock` gets initialized. Otherwise I feel like this is going to get really confusing down the line.

Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: Yeah it is. It turns out that if it's a `date` we actually fail hard on receiving the submission right now (because it parses to a `date` and later tries to save it to a `datetime` property). The other option was to coerce it to a `datetime` during submission and always make it a `date` in tests - which I guess now that you mention it could be better. We would just need to make sure that we add a test for the situation of submitting it as a `datetime` and restoring it as a `date`.

@mkangia we can look at this today.

File: corehq/apps/ota/tests/test_search_claim_endpoints.py [hunk 5]
```diff
@@ -137,4 +138,4 @@ def test_search_endpoint(self):
         url = reverse('sync_search', kwargs={'domain': DOMAIN})
         response = client.get(url, {'name': 'Jamie Hand', 'case_type': CASE_TYPE})
         self.assertEqual(response.status_code, 200)
-        self.assertEqual(re.sub(PATTERN, TIMESTAMP, response.content), known_result)
+        self.assertEqual(re.sub(DATE_PATTERN, FIXED_DATESTAMP, re.sub(PATTERN, TIMESTAMP, response.content)), known_result)
```
Comment (dimagimon) [inline corehq/apps/ota/tests/test_search_claim_endpoints.py]: E501 line too long (123 > 115 characters)

File: corehq/ex-submodules/casexml/apps/case/mock.py [hunk 6]
```diff
@@ -99,7 +99,10 @@ def __init__(self,
         """
         super(CaseBlock, self).__init__()
         self._id = case_id
-        date_modified = date_modified or datetime.utcnow()
+        now = datetime.utcnow()
+        date_modified = date_modified or now
+        date_opened = date_opened or (date_modified.date()\
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/mock.py]: E502 the backslash is redundant between brackets

**Comment by mkangia** [comment]
@czue : Finally a green build. Please re-review.

File: corehq/form_processor/backends/couch/update_strategy.py [hunk 7]
```diff
@@ -20,6 +21,20 @@
 from dimagi.ext.couchdbkit import StringProperty
 
 
+def _coerce_to_datetime(v):
```
Comment (proteusvacuum) [inline corehq/form_processor/backends/couch/update_strategy.py]: does this return `None` if `v` is already a `datetime`? I would potentially either assert that you are only calling this on `date` items, or return the input if it is already a `datetime`.

**Comment by proteusvacuum** [comment]
Awesome! No blockers from me (just one question).
Have you tested that this works properly with a sync on mobile? Excited to get this in!

File: corehq/ex-submodules/casexml/apps/case/tests/test_from_xform.py [hunk 8]
```diff
@@ -9,7 +9,7 @@
 from corehq.form_processor.interfaces.processor import FormProcessorInterface
 from corehq.form_processor.models import CaseTransaction, CommCareCaseSQL
 from corehq.form_processor.tests.utils import run_with_all_backends
-
+from corehq.form_processor.backends.couch.update_strategy import _coerce_to_datetime
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/tests/test_from_xform.py]: it might be worth moving `_coerce_to_datetime` and removing the leading `_` if it's going to be used as a utility in tests

File: corehq/form_processor/tests/test_basic_cases.py [hunk 9]
```diff
@@ -48,7 +48,7 @@ def test_create_case(self):
         self.assertEqual(case.owner_id, 'owner1')
         self.assertEqual(case.type, 'demo')
         self.assertEqual(case.name, 'create_case')
-        self.assertEqual(case.opened_on, modified_on)
+        self.assertEqual(case.opened_on.date(), modified_on.date())
```
Comment (czue) [inline corehq/form_processor/tests/test_basic_cases.py]: why was this change necessary?

File: corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py [hunk 10]
```diff
@@ -109,7 +109,7 @@ def test_couch_soft_rebuild(self):
         self.assertEqual(case.p3, 'p3-2') # new
         self.assertEqual(4, len(case.actions)) # create + update (2 actions) + 2 updates
         a0 = case.actions[1]
-        self.assertEqual(a0.updated_known_properties['opened_on'], case.opened_on)
+        self.assertEqual(a0.updated_known_properties['opened_on'], case.opened_on.date())
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py]: i'm actually confused by this - shouldn't the update / save still save a datetime? i thought we only converted to dates during restore

File: corehq/ex-submodules/couchforms/tests/test_edits.py [hunk 11]
```diff
@@ -274,7 +274,7 @@ def test_case_management_ordering(self):
 
         if not settings.TESTS_SHOULD_USE_SQL_BACKEND:
             self.assertEqual(
-                [create_form_id, edit_form_id, second_edit_form_id],
+                [create_form_id, create_form_id, edit_form_id, second_edit_form_id],
```
Comment (czue) [inline corehq/ex-submodules/couchforms/tests/test_edits.py]: can you clarify this one too?

**Comment by czue** [comment]
the other thing I was thinking was that we should add a test that explicitly submits date_opened as a datetime and tests that it restores as a date (if we don't have that yet). let me know if you want more input on what I think that would look like.

File: corehq/form_processor/backends/couch/update_strategy.py (hunk 7 above)
Comment (mkangia) [inline corehq/form_processor/backends/couch/update_strategy.py]: Yup. that needs a tweak. Thanks Farid.

File: corehq/ex-submodules/casexml/apps/case/tests/test_from_xform.py (hunk 8 above)
Comment (mkangia) [inline corehq/ex-submodules/casexml/apps/case/tests/test_from_xform.py]: Renamed.

File: corehq/ex-submodules/couchforms/tests/test_edits.py (hunk 11 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/tests/test_edits.py]: As found out with Cory, this is needed since date_opened is added now which results in an additional update block on the create_form causing create_form_id to repeat here.
@czue : is that quoted correctly?

**Comment by mkangia** [comment]
@proteusvacuum : Yup did test in once. Saw it breaking before the fix and working fine after the fix. Just waiting for a green build. Will check it again after that for a final confirmation.

File: corehq/ex-submodules/couchforms/tests/test_edits.py (hunk 11 above)
Comment (czue) [inline corehq/ex-submodules/couchforms/tests/test_edits.py]: :+1: yep

**Comment by mkangia** [comment]
@proteusvacuum @czue : Tested once again locally. Does not break much.
Tested by deleting all user data. Syncing again and then tried submitting a new case.
Previously it used to fail on accessing app post Re-syncing.

**Comment by proteusvacuum** [comment]
@mkangia Sounds good! There's merge conflicts with master right now.

**Comment by czue** [comment]
@mkangia is this ready to go? label still on

**Comment by mkangia** [comment]
@czue : I did test it once and all checks have passed. Removing the label. Hope this goes down well 😟

**Comment by czue** [comment]
@proteusvacuum feeling good about this? i'm ready to press the button but wanted to be safe.

**Comment by proteusvacuum** [comment]
Also tested locally. Seems g2g! 
Just as a sanity check - in this PR, the `date_opened` element is also added to V1 restores. I think these are completely unused, but wanted to make sure that this was intentional and / or won't break anything that I can't test.

**Comment by proteusvacuum** [comment]
(Also: didn't obviously see a resolution to this in the commits that came after it:

> the other thing I was thinking was that we should add a test that explicitly submits date_opened as a datetime and tests that it restores as a date (if we don't have that yet). let me know if you want more input on what I think that would look like. 

Not sure if you guys offlined or found that a test like this already existed)

**Comment by mkangia** [comment]
@proteusvacuum here is the commit for the test case added https://github.com/dimagi/commcare-hq/pull/12690/commits/56e827427f4aa7d34124a7874fd241c66273ab0c to ensure the coercion.
Also, for `the date_opened element is also added to V1 restores`, i did try to stick to just test cases and not any application changes. What change are you referring to?

**Comment by proteusvacuum** [comment]
@mkangia, thanks for pointing out the test case, looks good. 
For the V1 restore, I'm referring to these lines: https://github.com/dimagi/commcare-hq/blob/c1f859b8f6b53b89203eaed51cb98b19b0710c81/corehq/ex-submodules/casexml/apps/case/xml/generator.py#L126-L127, which got added here: https://github.com/dimagi/commcare-hq/commit/62d702bfea7731adf62c4b0d4a0deb7536be52b2

**Comment by mkangia** [comment]
@proteusvacuum i am not fully aware of the context of this. I think i attempted to just sync V1 with changes. Is this needed just for V2? I don't mind reverting the changes for V1 if thats not needed. Would there be a way to test it?

**Comment by proteusvacuum** [comment]
I just checked and I think we no longer support `V1`, so I don't think it matters.

**Comment by mkangia** [comment]
@proteusvaccum cool.

**Comment by czue** [comment]
:fearful: 

@NoahCarnahan just to be aware this is the change that broke syncs last go-around so probably good to do some monitoring after deploy. we were pretty diligent about testing but just wanted to put on your radar.

**Comment by mkangia** [comment]
🎆 ✌️

**Comment by NoahCarnahan** [comment]
Thanks for the heads up

**Comment by proteusvacuum** [comment]
🎉 🙈
